File: YapBook/Backend/Users/userController.js
const userService = require('./userService');

exports.getAllUsers = async (req, res) => {
    try {
        const allUsers = await userService.getAllUsers(req.user);
        res.status(200).json(allUsers);
    } catch (error) {
        res.status(500).json({ message: 'Something went wrong, try again' });
    }
};

exports.searchUser = async (req, res) => {
    try {
        const searchTerm = req.query.searchTerm;
        const users = await userService.searchUser(searchTerm);
        res.status(200).json(users);
    } catch (error) {
        res.status(500).json({ message: 'Something went wrong, try again' });
    }
};

exports.getUser = async (req, res) => {
    try {
        const userId = parseInt(req.params.id);
        const user = await userService.getUser(userId);
        res.status(200).json(user);
    } catch (error) {
        res.status(500).json({ message: 'Internal server error, could not get user' });
    }
};

exports.register = async (req, res) => {
    try {
        const newUser = await userService.register(req.body);
        res.status(200).json({ message: 'Registered successfully, please log in', user: newUser });
    } catch (error) {
        res.status(500).json({ message: 'Internal server error, could not create user' });
    }
};

exports.guestLogin = async (req, res) => {
    try {
        const token = await userService.guestLogin();
        res.status(200).json({ message: 'Guest login successful', token });
    } catch (error) {
        res.status(500).send('Something went wrong, try again');
    }
};

exports.login = async(req, res) => {
    userService.login(req, res);
};

exports.logout = async (req, res) => {
    try {
        await userService.logout(req.headers.authorization);
        res.status(200).json({ message: 'Logout successful' });
    } catch (error) {
        res.status(500).json({ message: 'Internal server error, error' });
    }
};

exports.editPhoto = async (req, res) => {
    try {
        const updatedProfile = await userService.editPhoto(req.user, req.file);
        res.status(200).json({ message: 'Profile updated', user: updatedProfile });
    } catch (error) {
        res.status(500).json({ message: 'Something went wrong, try again' });
    }
};

exports.deletePhoto = async (req, res) => {
    try {
        await userService.deletePhoto(req.user);
        res.status(200).json({ message: 'Image deleted' });
    } catch (error) {
        res.status(500).json({ message: 'Something went wrong, try again' });
    }
};

exports.editProfile = async (req, res) => {
    try {
        const updatedProfile = await userService.editProfile(req.user, req.body.bio);
        res.status(200).json({ message: 'Profile updated', user: updatedProfile });
    } catch (error) {
        res.status(500).json({ message: 'Something went wrong, try again' });
    }
};

exports.getFollowers = async (req, res) => {
    try {
        const followers = await userService.getFollowers(parseInt(req.params.id));
        res.status(200).json({ followers });
    } catch (error) {
        console.error('Error getting followers:', error);
        res.status(500).json({ message: 'Something went wrong, try again' });
    }
};

exports.getFollowing = async (req, res) => {
    try {
        const following = await userService.getFollowing(parseInt(req.params.id));
        res.status(200).json({ following });
    } catch (error) {
        console.error('Error getting following:', error);
        res.status(500).json({ message: 'Something went wrong, try again' });
    }
};

exports.follow = async (req, res) => {
    try {
        if (!req.user) {
            return res.status(401).json({ message: 'Unauthorized' });
        }

        const followerId = req.user.id;
        const followedId = parseInt(req.params.id, 10);

        if (isNaN(followedId) || isNaN(followerId)) {
            return res.status(400).json({ message: 'Invalid ID' });
        }

        const result = await userService.follow(followerId, followedId);
        return res.status(200).json(result); 
    } catch (error) {
        console.error('Error following/unfollowing user:', error);
        res.status(500).json({ message: 'Something went wrong, try again' });
    }
};

File: src/Backend-Java/src/main/java/com/yapbook/users/UserController.java
package com.yapbook.users;

import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.multipart.MultipartFile;
import org.springframework.web.bind.annotation.*;
import com.yapbook.security.JwtTokenUtil;

import java.util.HashMap;
import java.util.UUID;
import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/users")
public class UserController {
    @Autowired
    private UserService userService;

    @Autowired
    private PasswordEncoder passwordEncoder;

    @Autowired
    private JwtTokenUtil jwtTokenUtil;

    @GetMapping("/")
    public ResponseEntity<List<UserDTO>> getAllUsers(@RequestHeader("Authorization") String authToken) {
        try {
            String token = authToken.replace("Bearer ", "");
            Integer userId = jwtTokenUtil.getUserIdFromToken(token); 

            List<UserDTO> users = userService.getAllUsers(userId);
            return ResponseEntity.ok(users);
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(null);
        }
    }

    @GetMapping("/search")
    public ResponseEntity<?> searchUser(@RequestParam("searchTerm") String searchTerm) {
        try {
            List<UserDTO> users = userService.searchUser(searchTerm);
            return ResponseEntity.ok(users);
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body("Something went wrong, try again");
        }
    }

    @GetMapping("/{id}")
    public ResponseEntity<?> getUser(@PathVariable Integer id) {
        try {
            UserDTO userDTO = userService.getUser(id);
            return ResponseEntity.ok(userDTO);
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body("Internal server error, could not get user");
        }
    }

    @PostMapping("/register")
    public ResponseEntity<?> register(@RequestBody Map<String, String> request) {
        try {
            String email = request.get("email");
            String username = request.get("username");
            String password = request.get("password");
            String confirmPassword = request.get("confirmPassword");

            UserDTO newUserDTO = userService.register(email, username, password, confirmPassword);

            Map<String, Object> map = new HashMap<>();
            map.put("message", "Registered successfully, please log in." );
            map.put("user", newUserDTO);

            return ResponseEntity.ok(map);
        } catch (IllegalArgumentException e) {
            return ResponseEntity.status(HttpStatus.BAD_REQUEST).body("Passwords do not match");
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body("Internal server error, could not create user");
        }
    }

  @PostMapping("/login")
  public ResponseEntity<Map<String, Object>> login(@RequestBody Map<String, Object> request) {
          String username = (String) request.get("username");
          String password = (String) request.get("password");

          Map<String, Object> response = userService.login(username, password);

          return ResponseEntity.ok(response);
  }

  @PostMapping("/guest")
  public ResponseEntity<?> guestLogin() {
      try {
          String token = userService.guestLogin();

          return ResponseEntity.ok(Map.of("message", "Guest login successful", "token", token));
      } catch (Exception e) {
          return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body("Something went wrong, try again");
      }
  }

    @PostMapping("/logout")
    public ResponseEntity<?> logout(@RequestHeader("Authorization") String authorization) {
        try {
            String token = authorization.substring(7);
            userService.logout(token);
            return ResponseEntity.ok("Logout successful");
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body("Internal server error, error");
        }
    }

    @PostMapping("/edit-photo")
    public ResponseEntity<?> editPhoto(@RequestAttribute("user") User currentUser, @RequestParam("file") MultipartFile file) {
        try {
            UserDTO updatedProfile = userService.editPhoto(currentUser, file);
            return ResponseEntity.ok("Profile updated successfully, New Profile: " + updatedProfile);
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body("Something went wrong, try again");
        }
    }

    @PostMapping("/delete-photo")
    public ResponseEntity<?> deletePhoto(@RequestAttribute("user") User currentUser) {
        try {
            userService.deletePhoto(currentUser);
            return ResponseEntity.ok("Image deleted successfully");
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body("Something went wrong, try again");
        }
    }

    @PostMapping("/edit-profile")
    public ResponseEntity<?> editProfile(@RequestAttribute("user") User currentUser, @RequestBody String bio) {
        try {
            var updatedProfile = userService.editProfile(currentUser, bio);
            return ResponseEntity.ok("Profile updated successfully, New Bio: " + updatedProfile);
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body("Something went wrong, try again");
        }
    }

    @GetMapping("/{id}/followers")
    public ResponseEntity<?> getFollowers(@PathVariable Integer id) {
        try {
            List<UserDTO> followers = userService.getFollowers(id);
            return ResponseEntity.ok(followers);
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body("Something went wrong, try again");
        }
    }

    @GetMapping("/{id}/following")
    public ResponseEntity<?> getFollowing(@PathVariable Integer id) {
        try {
            List<UserDTO> following = userService.getFollowing(id);
            return ResponseEntity.ok(following);
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body("Something went wrong, try again");
        }
    }

    @PostMapping("/{id}/follow")
    public ResponseEntity<?> follow(@RequestHeader("Authorization") String authToken, @PathVariable Integer id) {
        try {
            String token = authToken.replace("Bearer ", "");
            Integer currentUserId = jwtTokenUtil.getUserIdFromToken(token);

            String result = userService.follow(currentUserId, id);
            return ResponseEntity.ok(result);
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body("Something went wrong, try again");
        }
    }
}
